# AI Workforce Intelligence Platform
# Notebook: 01_data_understanding.ipynb
# Author: Joshua Fernandes

# Objective

The objective of this notebook is to develop a comprehensive understanding of the IBM HR Analytics Employee Attrition dataset before any preprocessing or model development.

This phase includes:

- Dataset exploration
- Data structure analysis
- Data type verification
- Missing value analysis
- Duplicate record detection
- Constant feature identification
- Unique value analysis
- Descriptive statistics
- Categorical feature exploration
- Initial outlier detection

The insights obtained from this notebook will guide feature engineering, machine learning model development, and business intelligence reporting.

In [1]:
import warnings

import numpy as np
import pandas as pd

warnings.filterwarnings("ignore")

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)
pd.set_option("display.float_format", "{:.2f}".format)

In [2]:
df = pd.read_csv("..//data//raw//HR-Employee-Attrition.csv")

In [3]:
print("First Five Records")
display(df.head())

print("\nLast Five Records")
display(df.tail())

First Five Records


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,Sales Executive,4,Single,5993,19479,8,Y,Yes,11,3,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,Research Scientist,2,Married,5130,24907,1,Y,No,23,4,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,Laboratory Technician,3,Single,2090,2396,6,Y,Yes,15,3,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,Research Scientist,3,Married,2909,23159,1,Y,Yes,11,3,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,Laboratory Technician,2,Married,3468,16632,9,Y,No,12,3,4,80,1,6,3,3,2,2,2,2



Last Five Records


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
1465,36,No,Travel_Frequently,884,Research & Development,23,2,Medical,1,2061,3,Male,41,4,2,Laboratory Technician,4,Married,2571,12290,4,Y,No,17,3,3,80,1,17,3,3,5,2,0,3
1466,39,No,Travel_Rarely,613,Research & Development,6,1,Medical,1,2062,4,Male,42,2,3,Healthcare Representative,1,Married,9991,21457,4,Y,No,15,3,1,80,1,9,5,3,7,7,1,7
1467,27,No,Travel_Rarely,155,Research & Development,4,3,Life Sciences,1,2064,2,Male,87,4,2,Manufacturing Director,2,Married,6142,5174,1,Y,Yes,20,4,2,80,1,6,0,3,6,2,0,3
1468,49,No,Travel_Frequently,1023,Sales,2,3,Medical,1,2065,4,Male,63,2,2,Sales Executive,2,Married,5390,13243,2,Y,No,14,3,4,80,0,17,3,2,9,6,0,8
1469,34,No,Travel_Rarely,628,Research & Development,8,3,Medical,1,2068,2,Male,82,4,2,Laboratory Technician,3,Married,4404,10228,2,Y,No,12,3,1,80,0,6,3,4,4,3,1,2


In [4]:
rows, columns = df.shape

print(f"Rows    : {rows}")
print(f"Columns : {columns}")

Rows    : 1470
Columns : 35


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [6]:
pd.DataFrame({
    "Column Name": df.columns
})

,Column Name
0,Age
1,Attrition
2,BusinessTravel
3,DailyRate
4,Department
5,DistanceFromHome
6,Education
7,EducationField
8,EmployeeCount
9,EmployeeNumber


In [7]:
pd.DataFrame({
    "Data Type": df.dtypes
})

,Data Type
Age,int64
Attrition,object
BusinessTravel,object
DailyRate,int64
Department,object
DistanceFromHome,int64
Education,int64
EducationField,object
EmployeeCount,int64
EmployeeNumber,int64


In [8]:
missing = (
    df.isnull()
      .sum()
      .to_frame("Missing Values")
)

missing["Percentage"] = (
    missing["Missing Values"] / len(df) * 100
)

missing.sort_values(
    by="Missing Values",
    ascending=False,
    inplace=True
)

missing

,Missing Values,Percentage
Age,0,0.00
Attrition,0,0.00
BusinessTravel,0,0.00
DailyRate,0,0.00
Department,0,0.00
DistanceFromHome,0,0.00
Education,0,0.00
EducationField,0,0.00
EmployeeCount,0,0.00
EmployeeNumber,0,0.00


In [9]:
duplicates = df.duplicated().sum()

print(f"Duplicate Rows: {duplicates}")

Duplicate Rows: 0


In [10]:
constant_features = [
    column
    for column in df.columns
    if df[column].nunique() == 1
]

constant_features

['EmployeeCount', 'Over18', 'StandardHours']

In [11]:
unique_values = pd.DataFrame({
    "Unique Values": df.nunique()
})

unique_values.sort_values(
    by="Unique Values",
    inplace=True
)

unique_values

,Unique Values
EmployeeCount,1
Over18,1
StandardHours,1
Attrition,2
OverTime,2
PerformanceRating,2
Gender,2
BusinessTravel,3
Department,3
MaritalStatus,3


In [12]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Age,1470.00,36.92,9.14,18.00,30.00,36.00,43.00,60.00
DailyRate,1470.00,802.49,403.51,102.00,465.00,802.00,1157.00,1499.00
DistanceFromHome,1470.00,9.19,8.11,1.00,2.00,7.00,14.00,29.00
Education,1470.00,2.91,1.02,1.00,2.00,3.00,4.00,5.00
EmployeeCount,1470.00,1.00,0.00,1.00,1.00,1.00,1.00,1.00
EmployeeNumber,1470.00,1024.87,602.02,1.00,491.25,1020.50,1555.75,2068.00
EnvironmentSatisfaction,1470.00,2.72,1.09,1.00,2.00,3.00,4.00,4.00
HourlyRate,1470.00,65.89,20.33,30.00,48.00,66.00,83.75,100.00
JobInvolvement,1470.00,2.73,0.71,1.00,2.00,3.00,3.00,4.00
JobLevel,1470.00,2.06,1.11,1.00,1.00,2.00,3.00,5.00


In [13]:
df.describe(include="object").T

,count,unique,top,freq
Attrition,1470,2,No,1233
BusinessTravel,1470,3,Travel_Rarely,1043
Department,1470,3,Research & Development,961
EducationField,1470,6,Life Sciences,606
Gender,1470,2,Male,882
JobRole,1470,9,Sales Executive,326
MaritalStatus,1470,3,Married,673
Over18,1470,1,Y,1470
OverTime,1470,2,No,1054


In [14]:
numerical_features = df.select_dtypes(
    include=["int64", "float64"]
).columns.tolist()

categorical_features = df.select_dtypes(
    include="object"
).columns.tolist()

print("Numerical Features")
print(numerical_features)

print("\nCategorical Features")
print(categorical_features)

Numerical Features
['Age', 'DailyRate', 'DistanceFromHome', 'Education', 'EmployeeCount', 'EmployeeNumber', 'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked', 'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']

Categorical Features
['Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus', 'Over18', 'OverTime']


In [15]:
feature_categories = {
    "Target": [
        "Attrition"
    ],

    "Demographics": [
        "Age",
        "Gender",
        "MaritalStatus",
        "Education",
        "EducationField",
        "DistanceFromHome"
    ],

    "Job Information": [
        "Department",
        "JobRole",
        "JobLevel",
        "BusinessTravel"
    ],

    "Compensation": [
        "MonthlyIncome",
        "DailyRate",
        "HourlyRate",
        "MonthlyRate",
        "PercentSalaryHike",
        "StockOptionLevel"
    ],

    "Experience": [
        "TotalWorkingYears",
        "YearsAtCompany",
        "YearsInCurrentRole",
        "YearsSinceLastPromotion",
        "YearsWithCurrManager",
        "NumCompaniesWorked"
    ],

    "Performance": [
        "PerformanceRating",
        "TrainingTimesLastYear",
        "JobInvolvement"
    ],

    "Satisfaction": [
        "JobSatisfaction",
        "EnvironmentSatisfaction",
        "RelationshipSatisfaction",
        "WorkLifeBalance"
    ],

    "Administrative": [
        "EmployeeNumber",
        "EmployeeCount",
        "StandardHours",
        "Over18"
    ]
}

feature_categories

{'Target': ['Attrition'],
 'Demographics': ['Age',
  'Gender',
  'MaritalStatus',
  'Education',
  'EducationField',
  'DistanceFromHome'],
 'Job Information': ['Department', 'JobRole', 'JobLevel', 'BusinessTravel'],
 'Compensation': ['MonthlyIncome',
  'DailyRate',
  'HourlyRate',
  'MonthlyRate',
  'PercentSalaryHike',
  'StockOptionLevel'],
 'Experience': ['TotalWorkingYears',
  'YearsAtCompany',
  'YearsInCurrentRole',
  'YearsSinceLastPromotion',
  'YearsWithCurrManager',
  'NumCompaniesWorked'],
 'Performance': ['PerformanceRating',
  'TrainingTimesLastYear',
  'JobInvolvement'],
 'Satisfaction': ['JobSatisfaction',
  'EnvironmentSatisfaction',
  'RelationshipSatisfaction',
  'WorkLifeBalance'],
 'Administrative': ['EmployeeNumber',
  'EmployeeCount',
  'StandardHours',
  'Over18']}

In [16]:
from pathlib import Path

interim_path = Path("..//data//interim")
interim_path.mkdir(parents=True, exist_ok=True)

df.to_csv(interim_path / "hr_interim.csv", index=False)

print("Interim dataset saved successfully.")
print(f"Location: {interim_path / 'hr_interim.csv'}")

Interim dataset saved successfully.
Location: ..\data\interim\hr_interim.csv


# Key Findings
- The dataset contains 1,470 employee records and 35 features, providing a good foundation for HR analytics.
- No missing values or duplicate records were detected, indicating excellent data quality.
- Three constant features (EmployeeCount, Over18, and StandardHours) were identified and will be removed during preprocessing.
- EmployeeNumber is a unique identifier and will be excluded from model training.
- The target variable (Attrition) is moderately imbalanced, with approximately 84% of employees remaining and 16% leaving.
- The dataset includes a rich combination of demographic, compensation, experience, performance, and satisfaction features, enabling comprehensive exploratory analysis and predictive modeling.
- Several numerical features, such as MonthlyIncome, appear to exhibit skewed distributions and will be examined further during exploratory data analysis.
- Satisfaction-related variables provide an opportunity to engineer composite features such as Satisfaction Index, Burnout Score, and Retention Risk Score.